NIRISS updates

afeinstein20 opened the pull request:
Basically, making sure all the default NIRISS pieces are on the main branch before first stable release

kevin218 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -74,14 +85,12 @@ def read(filename, f277_filename, data, meta):
 
     # loads all the data into the data object
-    data.data = hdu['SCI',1].data + 0.0
+    data.data = hdu['SCI',1].data * hdu[0].header['EFFINTTM']
```
I don't think we want to be multiplying by the integration time just yet.  This step happens in `b2f.rate2count()`.

afeinstein20 replied:
removed

kevin218 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -74,14 +85,12 @@ def read(filename, f277_filename, data, meta):
-
-    data.var  = hdu['VAR_POISSON',1].data + 0.0
-    data.v0   = hdu['VAR_RNOISE' ,1].data + 0.0
+    data.var  = hdu['VAR_POISSON',1].data * hdu[0].header['EFFINTTM']**2.0
```
Do you make use of `data.var` in the code?  If so, we should include this variable in the conversion from rates to counts in `b2f.rate2count()`.

afeinstein20 replied:
removed, and yes I do! Although I think from the NIRISS meeting last week, it showed these might be overestimated... So I'll think about whether or not we really want to use these frames or not right now

kevin218 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -74,14 +85,12 @@ def read(filename, f277_filename, data, meta):
-    data.var  = hdu['VAR_POISSON',1].data + 0.0
-    data.v0   = hdu['VAR_RNOISE' ,1].data + 0.0
+    data.var  = hdu['VAR_POISSON',1].data * hdu[0].header['EFFINTTM']**2.0
+    data.v0   = hdu['VAR_RNOISE' ,1].data * hdu[0].header['EFFINTTM']**2.0
```
The code in `b2f.rate2count()` multiplies v0 by time (not time squared).  Can you double check that you really want time squared?

afeinstein20 replied:
removed for now

kevin218 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -460,53 +187,67 @@ def wave_NIRISS(wavefile, meta):
 
-    return meta
+    # Run assignwcs step on Stage 1 outputs:
+    assign_wcs_results = calwebb_spec2.assign_wcs_step.AssignWcsStep.call(rampfitting_results)
```
This step should have already happened in S2, so maybe we should also run `wave_NIRISS()` during S2 to calculate the wavelength solution?  Also, how does this differ from `hdulist['WAVELENGTH',1].data`?

afeinstein20 replied:
Unless I'm totally missing a step (which is possible!) this isn't an output from stage 2 as of right now

taylorbell57 replied:
Strange, this should definitely be happening in Stage 2 - it's a stage 2 function after all

afeinstein20 replied:
I found where this is happening now in Stage 2! Wavelengths are not saved in the calints file.

kevin218 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -460,53 +187,67 @@ def wave_NIRISS(wavefile, meta):
+    I think this is just a wrapper for fit_bg, because I perform outlier
+    flagging at the same time as the background fitting.
+    """
+    data, bkg, bkg_var = fit_bg(data, meta, readnoise, sigclip, 
```
These two steps (flag_bg and fit_bg) should be separate in order to coincide with the other instruments.  The latter needs to happen in `bg.BGsubtraction()`.

afeinstein20 replied:
This is fine for now as NIRISS has its own S3 reduction file. will need to reassess in next PR

kevin218 reviewed (commented):
I've reviewed about half of the files and thought I'd give you feedback now so that you can get started instead of waiting on me to finish.

kevin218 reviewed (commented):
Some more comments

afeinstein20 commented on eureka/S3_data_reduction/niriss_extraction.py:
```diff
@@ -0,0 +1,549 @@
+        col = medprof[:,i]+0.0
+        x = np.arange(0,len(col),1)
+
+        plt.plot(x, col, 'k')
```
that was an old testing plotting statement! it's been removed

afeinstein20 commented on eureka/S3_data_reduction/niriss_extraction.py:
```diff
@@ -0,0 +1,549 @@
+    cr_mask : np.ndarray , optional
+       A set of masks with cosmic rays already identified. This
+       will be used in the very last step, when extracting the
+       spectra. Default is None.
```
added

afeinstein20 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -0,0 +1,30 @@
+
+# Spectral extraction parameters
+which_table	2		# Which method used to find centers of orders
+proftype	gaussian	# What profilie shape to use
```
fixed

afeinstein20 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -0,0 +1,30 @@
+# Directories relative to project dir
+inputdir      	/Data/JWST-Sim/NIRISS/Stage2
+outputdir	/Data/JWST-Sim/NIRISS/Stage3
+ancildir	/Data/JWST-Sim/NIRISS/ancil
```
it's not used, I was just keeping with the generic directories for the other instruments. i've removed it now

afeinstein20 commented on eureka/S3_data_reduction/background.py:
```diff
@@ -416,23 +416,32 @@ def bkg_sub(img, mask, sigma=5, bkg_estimator='median',
+    background_error : np.ndarray
+       Error estimation on the background fitting.
     """
     sigma_clip = SigmaClip(sigma=sigma)
```
comments added!

afeinstein20 commented on eureka/S3_data_reduction/background.py:
```diff
@@ -416,23 +416,32 @@ def bkg_sub(img, mask, sigma=5, bkg_estimator='median',
     """
     sigma_clip = SigmaClip(sigma=sigma)
     
     if bkg_estimator.lower()=='mmmbackground':
```
definitely, and done!

afeinstein20 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -0,0 +1,30 @@
+suffix	  	 calints       # Data file suffix
+
+# Order location estimation
+method		 2		# Which method to identify orders by (see docs)
```
I've given these names (`edges` or `profile`) and have also changed the function names which call both of these methods to match (in `niriss.py`)

afeinstein20 replied:
this file has been updated

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -19,22 +21,28 @@
 from skimage import filters, feature
 from scipy.ndimage import gaussian_filter
 
+from jwst import datamodels
```
Right now, running S2 doesn't provide the wavelength maps for NIRISS, which is the only thing `jwst` is used for in S3. I'm happy to pop the `wave_NIRISS` function over there instead, but may need some guidance on the best way to do that

kevin218 replied:
Let's solve this problem after Exo IV.

afeinstein20 replied:
This has been resolved

afeinstein20 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -0,0 +1,30 @@
+# Spectral extraction parameters
+which_table	2		# Which method used to find centers of orders
+proftype	gaussian	# What profilie shape to use
+sigma		20		# Sigma outlier value
```
changed to `opt_sigma` since this is the sigma used in the optimal extraction routine?

afeinstein20 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -0,0 +1,30 @@
+# Diagnostics
+isplots_S3	3		# Generate a few (1), some (3), or many (5) diagnostic figures
+save_output	True		# Save outpurs for use in S4
+
```
added!

taylorbell57 commented on environment.yml:
```diff
@@ -16,6 +16,7 @@ dependencies:
     - george
     - gwcs
     - h5py
+    - image_registration
```
You shouldn't add image_registration. If your code is up-to-date with main, then image_registration should only be required if actually doing WFC3 analyses, in which case it can be installed using `pip install .[hst]` during installation. And we can't just do `pip install image_registration` or put it in the requirements.txt or environment.yml because the latest PyPI version is bugged and we need to use the newest version on GitHub, which is what the setup.py script does

taylorbell57 replied:
There was an old bug that meant that everyone needed to install image_registration, but the current version of the code shouldn't. Let me know if, after updating your branch against main, you still need to install image_registration - if so, I'll make sure that gets fixed

afeinstein20 replied:
Ah, yeah just following the "create a new environment" documentation, it wasn't working without `image_registration` so hopefully this is something that was fixed now

afeinstein20 replied:
I've removed it from the requirements.txt

afeinstein20 commented on eureka/S3_data_reduction/niriss_extraction.py:
```diff
@@ -0,0 +1,549 @@
+    m1, m2 = boxmask
+
+    for i in range(len(data)):
+        spec1[i] = np.nansum(data[i], # I think this needs to mask 2nd order?
```
I believe I've fixed this in the recent push!

afeinstein20 commented on eureka/S3_data_reduction/niriss_extraction.py:
```diff
@@ -0,0 +1,549 @@
+
+        for quad in range(1,4): # CHANGE BACK TO 4
+            # Figures out which quadrant location to use
+            if quad == 1: # Isolated first order (top right)
```
as is... no. over the next week I'm working on making all the algorithms flexible for both subarray sizes

afeinstein20 replied:
as is, I think it should work just fine? at least for the first order. I haven't had any test data to play with of the non-(2048x256) size, so I need to do that in the coming weeks.

kevin218 reviewed (commented):
I'm going to enable Xarrays on this branch over the next few days.

kevin218 commented on demos/JWST/S2_NIRISS_fs_template:
```diff
@@ -0,0 +1,41 @@
+
+# Stage 2 Documentation: https://eurekadocs.readthedocs.io/en/latest/ecf.html#stage-2
+
+suffix                  rateints    # Data file suffix
```
This should be `rampfitstep`

afeinstein20 replied:
fixed

kevin218 commented on demos/JWST/S2_NIRISS_fs_template:
```diff
@@ -0,0 +1,41 @@
+suffix                  rateints    # Data file suffix
+
+# Controls the cross-dispersion extraction
+slit_y_low              -1      # Use None to rely on the default parameters
```
These should both be `None`

afeinstein20 replied:
fixed!

kevin218 commented on demos/JWST/S2_NIRISS_fs_template:
```diff
@@ -0,0 +1,41 @@
+
+# Note: different instruments and modes will use different steps by default
+skip_bkg_subtract       False   # Not run for TSO observations
+skip_imprint_subtract   True    # Not run for NIRSpec Fixed Slit
```
Change comments on Lines 16-29 to reference NIRISS.
Double check to make sure these True/False settings are correct for NIRISS.

kevin218 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -4,25 +4,25 @@ ncpu	  	 1	       # Number of CPUs
 # Order location estimation
-method		 2		# Which method to identify orders by (see docs)
-save		 False		# Saves a table of central value for each order
+method		 2		    # Which method to identify orders by (see docs)
```
Don't these methods now have names?

afeinstein20 replied:
they do! I realized I changed it in 1 place, but not this one. It's been updated

kevin218 commented on eureka/S3_data_reduction/s3_reduce.py:
```diff
@@ -176,7 +176,8 @@ def reduce(eventlabel, ecf_path=None, s2_meta=None):
                              'you should edit the flag_bg() function in '
                              'nirspec.py and look at Issue #193 on Github!')
             elif meta.inst == 'niriss':
-                from . import niriss as inst
```
I'm pretty sure we want to import `niriss` here.

kevin218 commented on demos/JWST/S2_NIRISS_fs_template:
```diff
@@ -0,0 +1,41 @@
+slit_y_high             50      # Use None to rely on the default parameters
+
+# Modify the existing file to change the dispersion extraction - FIX: DOES NOT WORK CURRENTLY
+wave_orders         [1,2,3]     # Which orders to get wavelengths for
```
Code requires `waverange_start` and `waverange_end`.  Default should be None.

afeinstein20 replied:
added

kevin218 commented:
Time and units are stored in data. No need to save them in meta.

taylorbell57 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -0,0 +1,39 @@
+verbose         True        # If True, more details will be printed about steps
+
+# Project directory
+topdir		/Users/belugawhale/Documents/niriss_sim/tiny
```
Can you make this generic like the others ("topdir    /home/User/")

taylorbell57 commented:
Before this gets merged, can we make sure it is consistent with flake8, that all the documentation builds correctly, and that a test for Stage 3 is written (if one doesn't exist yet).

For flake8, you can use the following command (I'll add these flake8 settings to the setup.cfg file once Sebastian restructures Eureka. I'm also looking into having GitHub do this automatically for us). Run this in the main Eureka folder:
```
flake8 --exclude="./build/**/*,./docs/**/*,.vscode/*,**/site-packages/**/*,*.c,*.pyx,**/*niriss*.py" --per-file-ignores="**/__init__.py:F401,**/manageevent.py:F841,./demos/**/run_*.py:E402,**/test_*.py:E402,**/*.py:E722,**/readECF.py:F401,**/readECF.py:E722" .
```
The following messages exist for the main branch and don't need to be resolved by you:
```
./eureka/S3_data_reduction/background.py:338:41: F821 undefined name 'mask2'
./tests/test_MIRI.py:12:5: F401 'eureka.S1_detector_processing.s1_process as s1' imported but unused
./demos/JWST/run_eureka.py:5:1: F401 'eureka.S1_detector_processing.s1_process as s1' imported but unused
```

For the docs, run `make clean html` in the docs folder and fix any errors or warnings (none exist in the main branch).

kevin218 commented:
Can we halt any more commits to this PR and get it merged?

afeinstein20 commented:
yes!

taylorbell57 commented:
Should I review this tomorrow?

afeinstein20 commented:
@taylorbell57 yes please!

taylorbell57 commented on eureka/lib/tracing_niriss.py:
```diff
@@ -0,0 +1,410 @@
+    gcenters_2 = rm_outliers(gcenters_2)
+    x = np.arange(0,len(gcenters_1),1)
+
+    fit1 = clean_and_fit(x, x[x>800],
+                         f_centers, gcenters_1[x>800])
+    fit2 = clean_and_fit(x, x[(x>800) & (x<1800)],
+                         f_centers, gcenters_2[(x>800) & (x<1800)])
+
```
Should explain hardcoded values here too

taylorbell57 commented on eureka/lib/tracing_niriss.py:
```diff
@@ -0,0 +1,410 @@
+    """
+    There are some hard-coded numbers in here right now. The idea
+    is that once we know what the real data looks like, nobody will
+    have to actually call this function and we'll provide a CSV
```
CSV isn't great as it's hard to package together with the rest of the code, but I understand the general idea

taylorbell57 commented on eureka/lib/tracing_niriss.py:
```diff
@@ -0,0 +1,410 @@
+       Table with x,y positions for the first and second NIRISS
+       orders.
+    """
+    import matplotlib.pyplot as plt
```
Don't import things inside functions

taylorbell57 commented on eureka/lib/tracing_niriss.py:
```diff
@@ -0,0 +1,410 @@
+        plt.legend(ncol=3)
+        plt.show()
+
+    summed = data.median+0.0#np.nansum(data.median, axis=0)
```
Don't copy with adding zero. Also remove commented code if not needed anymore

taylorbell57 commented on eureka/lib/tracing_niriss.py:
```diff
@@ -0,0 +1,410 @@
+    summed = data.median+0.0#np.nansum(data.median, axis=0)
+    ccd = CCDData(summed*units.electron)
+
+    new_ccd_no_premask = ccdp.cosmicray_lacosmic(ccd, readnoise=150,
```
Likely best to come up with a shorter variable name

taylorbell57 commented on eureka/lib/util.py:
```diff
@@ -19,12 +19,22 @@ def readfiles(meta):
         data fits files.
     """
-    meta.segment_list = []
+    if suffix is None:
+        suffix = meta.suffix
+        return_meta=True
+    else:
+        return_meta=False
```
This is a really confusing way of changing what gets returned from this function. You should just change your code to accept the way this function was already setup and revert the changes to this function

taylorbell57 commented on eureka/niriss_class.py:
```diff
@@ -0,0 +1,516 @@
+import os
```
If this file is going to contain nothing other than Stage 3 code, it should be moved to S3_data_reduction

taylorbell57 replied:
Also, too much of this code is already written in other functions. You should be calling those other functions instead of duplicating code

taylorbell57 commented on eureka/niriss_class.py:
```diff
@@ -0,0 +1,516 @@
+        """
+        Sets up all proper attributes from the FITS file.
+
+        Attributes
```
The attributes of a class should be documented up above, before the init function

taylorbell57 commented on eureka/niriss_class.py:
```diff
@@ -0,0 +1,516 @@
+        median : np.ndarray
+           Median frame of all science images.
+        """
+        hdu = fits.open(os.path.join(self.data_dir, self.filename))
```
Most of this code is a copy of the read function and should just call that function instead of duplicating code (which makes it very challenging to maintain)

taylorbell57 commented on eureka/niriss_class.py:
```diff
@@ -0,0 +1,516 @@
+
+        return
+
+    def setup_f277(self, filename):
```
This is already a part of the niriss.read function

taylorbell57 commented on eureka/niriss_class.py:
```diff
@@ -0,0 +1,516 @@
+           - `self.var`
+        """
+        print('Cleaning data . . .')
+        self.data = interpolating_image(self.data, mask=self.dq)
```
Why is this never called in the s3_reduce_niriss way of running the code?

afeinstein20 replied:
oop because I forgot. added now

taylorbell57 commented on eureka/niriss_class.py:
```diff
@@ -0,0 +1,516 @@
+        """
+        hdu = fits.open(os.path.join(self.data_dir,
+                                     filename) )
+        self.f277 = hdu[1].data + 0.0
```
Don't copy by adding zero

afeinstein20 replied:
fixed (in upcoming push)

taylorbell57 commented on eureka/niriss_class.py:
```diff
@@ -0,0 +1,516 @@
+        """
+        wmap = wavelength(os.path.join(self.data_dir, self.filename),
+                          orders, inclass=True)
+        self.wavelength_map = wmap + 0.0
```
Don't copy by adding zero

taylorbell57 replied:
Can just do `self.wavelength_map = wavelength(...`

taylorbell57 commented on eureka/niriss_class.py:
```diff
@@ -0,0 +1,516 @@
+           trace x,y position values. Default is None. This is
+           a required parameter if you are running `method=='ref'`.
+
+        Attributes
+        ----------
+        trace_ear : astropy.table.Table
+           Astropy table with x,y coordinates for each order.
+           `trace_ear` is initialized when using the method `edges`.
+        trace_edge : astropy.table.Table
+           Astropy table with x,y coordinates for each order.
+           `trace_edge` is initialized when using the method `centers`.
```
If these are newly added attributes, they should be explained up at the top of the class, not here

afeinstein20 replied:
added

afeinstein20 commented on eureka/niriss_class.py:
```diff
@@ -0,0 +1,516 @@
+
+        s, v = box_extract(d, self.var, self.box_mask_separate)
+
+        self.box_var1     = v[0] + 0.0
+        self.box_var2     = v[1] + 0.0
+        self.box_var3     = v[2] + 0.0
+
+        self.box_spectra1 = s[0] + 0.0
+        self.box_spectra2 = s[1] + 0.0
+        self.box_spectra3 = s[2] + 0.0
```
fixed (in upcoming push)

afeinstein20 commented on demos/JWST/S2_NIRISS_fs_template:
```diff
@@ -0,0 +1,43 @@
+waverange_end       None        # Which wavelength to end extraction
+
+# Note: different instruments and modes will use different steps by default
+skip_bkg_subtract       False   # Not run for TSO observations
+skip_imprint_subtract   True
+skip_msa_flagging       True
+skip_extract_2d         True
+skip_srctype            True
+skip_master_background  True
+skip_wavecorr           True
+skip_flat_field         instruments
+skip_straylight         True
+skip_fringe             True
+skip_pathloss           True
+skip_barshadow          True
+skip_photom             True	  # Recommended to skip to get better uncertainties out of Stage 3.
+skip_resample           True    # Not run for TSO observations
+skip_cube_build         True
+skip_extract_1d         False
```
these are up to date

taylorbell57 replied:
Sorry, my point was that it's helpful if there's a comment on lines not relevant to NIRISS explaining that the setting is not important - for example, for MIRI there are comments saying `# Not run for MIRI LRS Slitless` for the steps that are never run for that instrument

afeinstein20 commented on eureka/S3_data_reduction/background.py:
```diff
@@ -494,40 +512,55 @@ def fitbg3(data, order_mask, readnoise=11, sigclip=[4, 2, 3], isplots=0):
                                          sigclip=sigclip[n])
             ccd = CCDData(m1.data*units.electron)
-            mask[m1.mask] += 1
+            mask[m1.mask] += True
```
I don't think it really matters here if it's = True or += True

afeinstein20 commented on eureka/S3_data_reduction/background.py:
```diff
@@ -494,40 +512,55 @@ def fitbg3(data, order_mask, readnoise=11, sigclip=[4, 2, 3], isplots=0):
-    data.bkg_removed = bkg_subbed
-
-    return data
+    return np.abs(bkg), bkg_var, rm_crs
```
There are some very very few cases where I've seen a few negative pixel values. This is just a safety precaution for those edge cases.

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -2,16 +2,17 @@
 #
 # Written by: Adina Feinstein
 # Last updated by: Adina Feinstein
-# Last updated date: February 7, 2022
+# Last updated date: April 16, 2022
 #
 ####################################
```
This must be a recent thing, since I copied this from previous files! I'll delete it

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -20,21 +21,22 @@
 
 
-def read(filename, f277_filename, data, meta):
+def read(filename, data, meta, f277_filename=None):
```
fixed

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -54,634 +56,163 @@ def read(filename, f277_filename, data, meta):
+    data.attrs['intend'] = 3 #data.attrs['mhdr']['INTEND']
+
+    if f277_filename is not None:
+        f277= fits.open(f277_filename)
```
fixed!

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -54,634 +56,163 @@ def read(filename, f277_filename, data, meta):
+        # This exception is (hopefully) only for simulated data
+        print("WARNING: INT_TIMES not found. Using EXPSTART and EXPEND in UTC.")
+        time = np.linspace(data.attrs['mhdr']['EXPSTART'],
+                                  data.attrs['mhdr']['EXPEND'],
+                                  int(data.attrs['NINTS']))
```
fixed!

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -54,634 +56,163 @@ def read(filename, f277_filename, data, meta):
-            p = p[p>40] # sometimes catches an upper edge that doesn't exist
-        
-        peaks[i][:len(p)] = p
+    sci[np.isnan(sci) == True] = 0
+    err[np.isnan(sci) == True] = 0
```
fixed!

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -54,634 +56,163 @@ def read(filename, f277_filename, data, meta):
 
 
-def fit_bg(data, meta, n_iters=3, readnoise=11, sigclip=[4,4,4], isplots=0):
+def fit_bg(data, meta, log,
+           readnoise=11, sigclip=[4,4,4],
+           box=(5,2), filter_size=(2,2),
+           bkg_estimator=['median'],
+           testing=False, isplots=0):
```
updated

afeinstein20 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -0,0 +1,39 @@
+save_table 	    False		 # Saves a table of central value for each order
+extract_order3  False    # Sets whether to extract the 3rd order or mask it
+radius          1        # Sets the radius for edge extraction method
+filter          4        # Sets the filter size for edge extraction method
```
done!

afeinstein20 commented on eureka/S3_data_reduction/s3_reduce_niriss.py:
```diff
@@ -0,0 +1,416 @@
+                     mute=(not meta.verbose))
+
+        data = inst.fit_bg(data, meta, log,
+                           readnoise=meta.readnoise,
+                           sigclip=meta.sigclip,
+                           box=meta.box,
+                           filter_size=meta.filter_size,
+                           bkg_estimator=meta.bkg_estimator,
+                           testing=meta.testing_S3,
+                           isplots=meta.isplots_S3)
```
So that it plays nicely with the class testing method!

afeinstein20 commented on eureka/S3_data_reduction/bright2flux.py:
```diff
@@ -92,13 +92,19 @@ def dn2electrons(data, meta):
+    else:
+        subgain = gain + 0.0
+        # Convert to electrons
+        data['flux'][:,1:,1:] *= subgain
+        data['err'][:,1:,1:]  *= subgain
+        data['v0'][:,1:,1:]   *= (subgain)**2  # FINDME: should this really be squared
```
The gain reference file has a different shape from the NIRISS data files... This may need to be changed later, but the for the files we do have this is the only way it works

afeinstein20 commented on eureka/S3_data_reduction/bright2flux.py:
```diff
@@ -92,13 +92,19 @@ def dn2electrons(data, meta):
-    data['flux'] *= subgain
-    data['err'] *= subgain
-    data['v0'] *= (subgain)**2  # FINDME: should this really be squared
+    if meta.inst.lower() != 'niriss':
```
This is mainly to get around the fact that the gain file is of a different shape (see next comment)

afeinstein20 commented on eureka/S3_data_reduction/background.py:
```diff
@@ -494,40 +512,55 @@ def fitbg3(data, order_mask, readnoise=11, sigclip=[4, 2, 3], isplots=0):
+       Errors on the fitted backgrouns.
     """
+    if inclass is False:
+        data = data.flux + 0.0
```
fixed

afeinstein20 commented:
> Before this gets merged, can we make sure it is consistent with flake8, that all the documentation builds correctly, and that a test for Stage 3 is written (if one doesn't exist yet).
> 
> For flake8, you can use the following command (I'll add these flake8 settings to the setup.cfg file once Sebastian restructures Eureka. I'm also looking into having GitHub do this automatically for us). Run this in the main Eureka folder:
> 
> ```
> flake8 --exclude="./build/**/*,./docs/**/*,.vscode/*,**/site-packages/**/*,*.c,*.pyx,**/*niriss*.py" --per-file-ignores="**/__init__.py:F401,**/manageevent.py:F841,./demos/**/run_*.py:E402,**/test_*.py:E402,**/*.py:E722,**/readECF.py:F401,**/readECF.py:E722" .
> ```
> 
> The following messages exist for the main branch and don't need to be resolved by you:
> 
> ```
> ./eureka/S3_data_reduction/background.py:338:41: F821 undefined name 'mask2'
> ./tests/test_MIRI.py:12:5: F401 'eureka.S1_detector_processing.s1_process as s1' imported but unused
> ./demos/JWST/run_eureka.py:5:1: F401 'eureka.S1_detector_processing.s1_process as s1' imported but unused
> ```
> 
> For the docs, run `make clean html` in the docs folder and fix any errors or warnings (none exist in the main branch).

I've gone through and fixed all of the `flake8` issues affiliated with the files I have changed. I have also run the docs and they look good: no errors

taylorbell57 commented on eureka/lib/masking.py:
```diff
@@ -0,0 +1,128 @@
+    """
+    def interpolate(d, m):
+        try:
+            m = m is False
```
m = ~m is an order of magnitude faster for booleans. Also note that this line will end up getting run for integer arrays without error and will give the exact opposite behaviour as expected. And the m = ~m suggestion won't work for integer arrays

afeinstein20 replied:
This isn't my function and I've never run it. I was making changes based on the `flake8` command (since there were a bunch not relevant to my stuff). I've updated it for now, but may want to double check with the original author if this will break things down the line.

taylorbell57 replied:
GitHub shows that this file didn't exist before this PR though - the GitHub history shows that you wrote this 2 months ago with commit e374f053b3b51ee89ba1397a3e9972cf8bf3ea47. Do we no longer need this function?

taylorbell57 commented on eureka/lib/simultaneous_order_fitting.py:
```diff
@@ -0,0 +1,142 @@
+        # std of gaussian profile
+        sigs = np.linspace(sig[0], sig[1], length)
+        # generates all possible combos
+        combos = np.array(list(itertools.product(*[As, Bs, sigs])))
```
Why convert to a list and then convert to an array? Also, why do `*[As,Bs,sigs]` instead of just `As, Bs, sigs`?

taylorbell57 commented on eureka/lib/simultaneous_order_fitting.py:
```diff
@@ -0,0 +1,142 @@
+                                                   [final_guess[1]],
+                                                   [final_guess[2]],
+                                                   pos1, pos2,
+                                                   return_together=False)
```
If this is the default behaviour, no need to pass this keyword argument

taylorbell57 reviewed (changes requested):
There's still several files with comments unaddressed - not sure if those just aren't pushed yet or if you missed them. Make sure to check in the "Conversation" tab on the PR and look for unresolved comments. When you addressed the flake8 comment, it hid a lot of the comments in the "Files Changed" tab, but they're still visible in the "Conversation" tab.

Also, I messed up my flake8 command, please use:
```
flake8 --select=E1,E2,E3,E4,E5,E7,W6,F --ignore=W504,W503,E704,E226,E722 --exclude=./build/**/*,**/site-packages/**/*,./docs/**/*,.vscode/*,**/run_eureka*.py --per-file-ignores=**/__init__.py:F401,**/manageevent.py:F841,./demos/**/run_*.py:E402,**/test_*.py:E402,**/readECF.py:F401,**/hst_scan.py:E241,**/test_MIRI.py:F401,E402 .
```
Once my under review PR gets merged, that'll be simplified down to just `flake8 .` and it'll also automatically be done for us by GitHub, but for now we need to use that unweildy command.

afeinstein20 commented on eureka/lib/clipping.py:
```diff
@@ -195,10 +196,47 @@ def gauss_removal(img, mask, linspace, where='bkg'):
+    if testing is False:
+        y = img.shape[1]
+    else:
+        y = 10
```
This is just if testing is True. It then only performs `clipping.time_removal` along 10 columns, instead of the whole image.

afeinstein20 commented on docs/source/ecf.rst:
```diff
@@ -297,8 +296,57 @@ topdir + time_file
+
+bkg_estimator
+'''''''''''''
+How to calculate and estimate the background over a given box size. Options include 'median', 'mean', and 'MMMBackground' (3*mean + 2*median). Should be of length `sigclip`.
```
It doesn't make sense to write documentation from the other package. I've included a link to this routine in photutils so it is clearer.

afeinstein20 commented on docs/source/ecf.rst:
```diff
@@ -297,8 +296,57 @@ topdir + time_file
+
+filter_size
+'''''''''''
+The filter size of pixels in (ny,nx) to filter over. A filter size of 1 or (1, 1) means no filtering. Should be of length `sigclip`.
```
It doesn't make sense to write documentation from the other package. I've included a link to this routine in photutils so it is clearer.

afeinstein20 commented:
> ```
> flake8 --select=E1,E2,E3,E4,E5,E7,W6,F --ignore=W504,W503,E704,E226,E722 --exclude=./build/**/*,**/site-packages/**/*,./docs/**/*,.vscode/*,**/run_eureka*.py --per-file-ignores=**/__init__.py:F401,**/manageevent.py:F841,./demos/**/run_*.py:E402,**/test_*.py:E402,**/readECF.py:F401,**/hst_scan.py:E241,**/test_MIRI.py:F401,E402 .
> ```

This command is not working for me. I am getting:
`zsh: no matches found: --exclude=./build/**/*,**/site-packages/**/*,./docs/**/*,.vscode/*,**/run_eureka*.py`

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -45,6 +40,9 @@ def read(filename, f277_filename, data, meta):
        Single filename to read. Should be a `.fits` file.
     data : object
        Data object in which the fits data will be stored.
+    meta : object
```
resolved (in upcoming push)

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -54,634 +52,189 @@ def read(filename, f277_filename, data, meta):
 
     Parameters
     ----------
     data : object
     meta : object
```
see above

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -54,634 +52,189 @@ def read(filename, f277_filename, data, meta):
+                              meta.trace_edge, booltype=True,
+                              return_together=True)
+    import matplotlib.pyplot as plt
+    plt.imshow(box_mask)
+    plt.show()
```
oops, yeah debugging plot!

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -54,634 +52,189 @@ def read(filename, f277_filename, data, meta):
+        box_mask = dirty_mask(data.medflux.values,
+                              meta.trace_edge, booltype=True,
+                              return_together=True)
+    import matplotlib.pyplot as plt
```
this was just for the debugging plot. it's been removed now

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -54,634 +52,189 @@ def read(filename, f277_filename, data, meta):
     Parameters
     ----------
-    data : object
     meta : object
```
updated

afeinstein20 commented on eureka/S3_data_reduction/niriss.py:
```diff
@@ -54,634 +52,189 @@ def read(filename, f277_filename, data, meta):
 
     Returns
     -------
     data : object
```
completed

afeinstein20 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -0,0 +1,39 @@
+suffix	  	 calints       # Data file suffix
+
+# Order location estimation
+trace_method		ears     # Which method to identify orders by (see docs)
```
I've added docs describing the differences between these trace methods now

afeinstein20 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -0,0 +1,39 @@
+filter          4        # Sets the filter size for edge extraction method
+
+# Background parameters
+readnoise	    18		                   # Estimation of the readnoise
```
In theory it should be in the file headers I believe, but right now the parameters is not there. So I'm leaving this in the ecf for now

afeinstein20 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -0,0 +1,39 @@
+filter_size	  ((2,1), (1,1))		       # Window size to apply the 2D low-resolution background map
+
+# Spectral extraction parameters
+which_table	      	 centers 	  # Which method used to find centers of orders
```
added to doc strings now

afeinstein20 commented on demos/JWST/S3_niriss_template.ecf:
```diff
@@ -0,0 +1,39 @@
+# Background parameters
+readnoise	    18		                   # Estimation of the readnoise
+sigclip	     	[4,4,4]	          	     # Iterations and sigma to find & remove outliers
+box	        	((10,2), (5,2))		       # Box size to average background values over
+bkg_estimator	['median', 'median']     # How to calculate and estimate the background over the box size
+filter_size	  ((2,1), (1,1))		       # Window size to apply the 2D low-resolution background map
```
these are now fully described in the docs

taylorbell57 commented:
> > ```
> > flake8 --select=E1,E2,E3,E4,E5,E7,W6,F --ignore=W504,W503,E704,E226,E722 --exclude=./build/**/*,**/site-packages/**/*,./docs/**/*,.vscode/*,**/run_eureka*.py --per-file-ignores=**/__init__.py:F401,**/manageevent.py:F841,./demos/**/run_*.py:E402,**/test_*.py:E402,**/readECF.py:F401,**/hst_scan.py:E241,**/test_MIRI.py:F401,E402 .
> > ```
> 
> This command is not working for me. I am getting: `zsh: no matches found: --exclude=./build/**/*,**/site-packages/**/*,./docs/**/*,.vscode/*,**/run_eureka*.py`

Strange, I'm not sure what went wrong there... My PR got merged into main just now, so if you merge main into your branch (which you'll need to do to resolve the conflict with background.py) then you'll be able to just do `flake8 .` at the top of the Eureka folder (aka inside Eureka, not eureka)

taylorbell57 reviewed (changes requested):
Alright @afeinstein20, I've summarized here the remaining issues that still need to be addressed before this PR can be merged.

1. In the NIRISS read function, can you put an if/else statement that'll work once we have real data instead of hardcoding intstart and intend
2. Add documentation for all of the niriss_cython functions
3. Relatedly, make sure that all your functions' inputs and outputs are documented. If you don't do it now, it'll be much harder to remember what to put later. There are many functions with missing or outdated Parameters and Returns docstrings.
4. In niriss_extraction.py, only catch the type of exception you expect. Just using `except` will catch other things like memory overflow that you don't want to be catching.
5. In niriss_extraction.py, is the "CHANGE BACK TO 4" comment still relevant
6. simultaneous_order_fitting.py, tracing_niriss.py, and niriss_class.py all seem too Stage 3 and NIRISS specific, and should be moved to the Stage 3 folder.
7. In simultaneous_order_fitting.py, why do you do `nonlocal profile` instead of just passing in profile as an argument?
8. In tracing_niriss.py, explain using code comments the meaning of the lines `g[g > 4] = 10000` and `edges[edges > 0] = 10` and the hardcoded cutoffs in the lines calling the `clean_and_fit` function
9. Too much of niriss_class.py has now been written in other files and is now duplicated. Instead of duplicating the code, you should be calling the already written functions (e.g reading in images should only be done with `niriss.read()`).

You'll also need to try running Stage 3 to make sure that none of the changes I made broke things. Relatedly, please write at least an S3 test for NIRISS (following the example of the other instruments) so that we can all make sure that the changes we make won't break things for NIRISS. I did confirm that the tests for the other instruments all pass now.

kevin218 commented:
As soon as we move to the development branch, I want to reorganize the NIRISS modules and merge the components into the main S3 reduction pipeline. This means that many of the remaining changes listed above will have to be revisited, so let's not do them now.

The priority right now is to resolve the remaining conflicts then merge this PR as is.

taylorbell57 commented:
Alright, I can open an issue tomorrow to keep track of the other issues I'd mentioned, but then only #10 and making sure that I didn't break any of the rest of the code for NIRISS need to be done before merging. Ideally writing a Stage 3 NIRISS test too (should be very easy hopefully)

taylorbell57 commented:
Alright, I've removed the call to the now removed `wavelength` function in niriss_class.py, I've fixed the missing `traces` variable in s3_reduce_niriss.py, and I've added a S2-3 test for NIRISS. The test I added allowed me to fix some other bugs with the NIRISS code and ECFs, but I'm now at the point where niriss.fig_bg fails with the following trace
```
/Users/tjbell1/Research/JWST/ERS/Eureka/eureka/S3_data_reduction/s3_reduce.py:77: in reduce
    s3_reduce_niriss.reduce(eventlabel,
/Users/tjbell1/Research/JWST/ERS/Eureka/eureka/S3_data_reduction/s3_reduce_niriss.py:241: in reduce
    data = inst.fit_bg(data, meta, readnoise=meta.readnoise,
/Users/tjbell1/Research/JWST/ERS/Eureka/eureka/S3_data_reduction/niriss.py:158: in fit_bg
    bkg, bkg_var, cr_mask = fitbg3(data, np.array(box_mask-1, dtype=bool),
/Users/tjbell1/Research/JWST/ERS/Eureka/eureka/S3_data_reduction/background.py:582: in fitbg3
    b1, b1_err = bkg_sub(rm_crs[i],
/Users/tjbell1/Research/JWST/ERS/Eureka/eureka/S3_data_reduction/background.py:481: in bkg_sub
    b = Background2D(img, box,
/Users/tjbell1/miniconda3/envs/eureka/lib/python3.9/site-packages/photutils/background/background_2d.py:371: in __init__
    self._calc_bkg_bkgrms()
/Users/tjbell1/miniconda3/envs/eureka/lib/python3.9/site-packages/photutils/background/background_2d.py:711: in _calc_bkg_bkgrms
    idx = self._select_meshes(data_sigclip)
ValueError: All meshes contain > 2.0 (10.0 percent per mesh) masked pixels.  Please check your data or decrease "exclude_percentile".
```

Have you encountered this before @afeinstein20? Can you try running the code on the files you have locally to see if it's just a bug with the test files (perhaps they're too small or something) or if it's a bug with the code itself

taylorbell57 commented:
FYI, we restructured eureka now, and I just merged those changes into this branch to help get this PR merged sooner.

kevin218 commented:
Yes, I'm pretty sure this is only an issue that comes up for the test data.

afeinstein20 commented:
> 

I have seen this before, but for some reason it's only with the smaller testing set. With the full dataset, this issue doesn't appear. I've yet to figure out why... and it seems to be the last thing failing the testing right now

taylorbell57 commented:
Good question, I'd like to run this branch locally with the NIRISS simulated data (full dataset, not tiny) to make sure it is only the test that's failing. If so, then perhaps we can comment out the test and merge this PR and work out the testing issue later? I'm a bit hesitant to do that, but I don't know what to do to get the test working

taylorbell57 commented:
This PR is still failing for me locally with the full dataset. Trying to run S3, I get the following outputs:

```
Starting Stage 3 Reduction

Input directory: /Volumes/DataDrive/WASP-39b_NIRISS/Stage2/S2_2022-06-21_WASP-39b_NIRISS_run1/
Output directory: /Volumes/DataDrive/WASP-39b_NIRISS/Stage3/S3_2022-07-05_WASP-39b_NIRISS_run2/
Copying S3 control file

Found 4 data file(s) ending in calints.fits
Reading file 1 of 4
  WARNING: INT_TIMES not found. Using EXPSTART and EXPEND in UTC.
  WARNING: Number of SCI integrations doesn't match  NINTS from header. Updating NINTS.
  Automatically getting reference files to convert units to electrons
  Converting from data numbers per second (DN/s) to electrons
  WARNING: ERR has 1449536 NaNs. Your subregion may be off the edge of the detector subarray.
Masking NaN region and continuing, but you should really stop and reconsider your choices.
  WARNING: V0 has 1449536 NaNs. Your subregion may be off the edge of the detector subarray.
Masking NaN region and continuing, but you should really stop and reconsider your choices.
  Performing background outlier rejection
Traceback (most recent call last):
  File "/Users/tjbell1/Research/JWST/ERS/Eureka/ecfs_NIRISS/run_eureka.ecf", line 24, in <module>
    s3_meta = s3.reduce(eventlabel)
  File "/Users/tjbell1/Research/JWST/ERS/Eureka/ecfs_NIRISS/../src/eureka/S3_data_reduction/s3_reduce_niriss.py", line 249, in reduce
    data = inst.fit_bg(data, meta, readnoise=meta.readnoise,
  File "/Users/tjbell1/Research/JWST/ERS/Eureka/ecfs_NIRISS/../src/eureka/S3_data_reduction/niriss.py", line 148, in fit_bg
    box_mask = dirty_mask(data.medflux.values,
  File "/Users/tjbell1/Research/JWST/ERS/Eureka/ecfs_NIRISS/../src/eureka/S3_data_reduction/niriss_extraction.py", line 136, in dirty_mask
    order3[:, i] = img[s:e, i]
ValueError: could not broadcast input array from shape (0,) into shape (60,)
```

My ECF settings are as follows:
```
# Eureka! Control File for Stage 3: Data Reduction of NIRISS

# Stage 3 Documentation: https://eurekadocs.readthedocs.io/en/latest/ecf.html#stage-3

ncpu            1           # Number of CPUs
suffix          calints     # Data file suffix

# Order location estimation
trace_method    ears        # Which method to identify orders by (see docs)
poly_order      4           # Degree polynomial to fit to trace ear extraction method
save_table      False       # Saves a table of central value for each order
extract_order3  False       # Sets whether to extract the 3rd order or mask it
radius          1           # Sets the radius for edge extraction method
filter          4           # Sets the filter size for edge extraction method

# Background parameters
readnoise       18                      # Estimation of the readnoise
sigclip         [4,4]                   # Iterations and sigma to find & remove outliers
box             ((10,2), (5,2))         # Box size to average background values over
bkg_estimator   ['median', 'median']    # How to calculate and estimate the background over the box size
filter_size     ((2,1), (1,1))          # Window size to apply the 2D low-resolution background map

# Spectral extraction parameters
which_table     centers     # Which method used to find centers of orders
proftype        median      # What profile shape to use
opt_sigma       20          # Sigma outlier value
per_quad        True        # Extracts orders in quadrants or not

# Diagnostics
isplots_S3      3           # Generate a few (1), some (3), or many (5) diagnostic figures
save_output     True        # Save outpurs for use in S4
testing_S3      False       # Boolean, set True to only use last file and generate select figures
hide_plots      False       # If True, plots will automatically be closed rather than popping up
verbose         True        # If True, more details will be printed about steps

# Project directory
# Project directory
topdir                  /Volumes/DataDrive/WASP-39b_NIRISS

# Directories relative to project dir
inputdir                Stage2
outputdir               Stage3
```

@afeinstein20, are any of my settings incorrect, or do you also get this error if you try to run the code on your machine?

taylorbell57 reviewed (changes requested):
Alright, I fixed a few issues I encountered with your new updates, but there are a few things that you'll need to fix

a. Please make sure your newest S3 ecf settings are documented in the docs/source/ecf.rst file
b. Please run Stage 3 yourself and make sure the code works, and let us know this PR is ready for review again once it does. I was able to fix some of the bugs that were causing S3 to crash, but there are still some that I can't fix (e.g. meta has no attribute boxsize1, boxsize2, and boxsize3 at line 277 of s3_reduce_niriss.py).

Also, please make sure you update the testing ecfs at tests/NIRISS_ecfs anytime you update the demo ecfs (I did that for you this time, but keep that in mind for the future).

taylorbell57 commented on src/eureka/S3_data_reduction/s3_reduce_niriss.py:
```diff
@@ -0,0 +1,399 @@
+        medapdata = np.median(data.flux, axis=0)
+
+        # creates mask for the traces
+        box_masks = niriss_extraction.dirty_mask(medapdata, meta.trace,
+                                                 boxsize1=meta.boxsize1,
+                                                 boxsize2=meta.boxsize2,
+                                                 boxsize3=meta.boxsize3,
+                                                 isplots=meta.isplots_S3)
```
These are the lines with the undefined meta parameters that are causing the test and my local runs to fail. Have you been running the class-based method locally, or did you just forget to include boxsize 1-3 in the S3 NIRISS template?

afeinstein20 replied:
I included the boxsizes in the S3 template! Sorry, I forgot to push that (since those files are ignored I can't do it through my typical way). Should be updated now!

afeinstein20 commented:
works for me. lmk when I should pull your changes to test again

taylorbell57 commented:
Alright, I now seem to be able to get the test to run all the way until plots_s3.lc_nodriftcorr is called when an error is thrown because optspec is a 3D array for NIRISS since there are three orders instead of just a 2D array for all the other instruments. I can write a quick patch to get that working if you want. In the meantime, please document all your S3 ecf settings in docs/source/ecf.rst - there are several settings missing there.

afeinstein20 commented:
sounds good! I'll get to that ecf.rst file now

afeinstein20 commented:
`ecf.rst` has been updated

taylorbell57 reviewed (changes requested):
Alright, with several edits I can now get the code to run from start to finish. However, there are definitely some issues that still need to be addressed before this PR can be merged.


Most importantly, data.optspec and data.opterr need to be set correctly during Stage 3. Previously they were only being initialized to zero, so moving on to Stage 4 was going to be impossible as none of the values from Stage 3 were being saved. I've done something quite hacky to get the data out of `optimal_extraction_routine` into data.optspec and data.opterr (to be able to debug the rest of the code), but I have definitely not done the right thing and significantly more work needs to be spent on this.


Second, an error ends up popping up for the fourth x1dints file with the following traceback, so I had to remove the fourth segment for now:
```
Traceback (most recent call last):
  File "/Users/tjbell1/Research/JWST/ERS/Eureka/ecfs_NIRISS/run_eureka.ecf", line 24, in <module>
    s3_meta = s3.reduce(eventlabel)
  File "/Users/tjbell1/Research/JWST/ERS/Eureka/ecfs_NIRISS/../src/eureka/S3_data_reduction/s3_reduce_niriss.py", line 190, in reduce
    soln = hdulist[ext].data['WAVELENGTH']
  File "/Users/tjbell1/miniconda3/envs/eureka/lib/python3.9/site-packages/astropy/io/fits/fitsrec.py", line 501, in __getitem__
    return self.field(key)
  File "/Users/tjbell1/miniconda3/envs/eureka/lib/python3.9/site-packages/astropy/io/fits/fitsrec.py", line 688, in field
    column = self.columns[key]
  File "/Users/tjbell1/miniconda3/envs/eureka/lib/python3.9/site-packages/astropy/io/fits/column.py", line 1644, in __getitem__
    key = _get_index(self.names, key)
  File "/Users/tjbell1/miniconda3/envs/eureka/lib/python3.9/site-packages/astropy/io/fits/column.py", line 2055, in _get_index
    raise KeyError(f"Key '{key}' does not exist.")
KeyError: "Key 'WAVELENGTH' does not exist."
```


Third, it seems there may be a bug with the optimal spectral extraction as my 2D light curves for each quadrant look horrible. Below is the 2D light curve for the first quadrant spectrum - the white space and incorrect x-axis is expected given my aforementioned hacking, but the red/blue mess is not.
![fig3101_quad0-2D_LC](https://user-images.githubusercontent.com/24514122/178385047-6a51e1b8-9243-4690-9e11-d8443281c070.png)


Also, if the following error is a simple bug for you to fix, please address it; if not, let's open this as an issue and address it once we get this PR merged. The NIRISS optimal spectral extraction fails for `per_quad=False`, giving the following traceback:
```
Traceback (most recent call last):
  File "/Users/tjbell1/Research/JWST/ERS/Eureka/ecfs_NIRISS/run_eureka.ecf", line 24, in <module>
    s3_meta = s3.reduce(eventlabel)
  File "/Users/tjbell1/Research/JWST/ERS/Eureka/ecfs_NIRISS/../src/eureka/S3_data_reduction/s3_reduce_niriss.py", line 296, in reduce
    optimal_extraction_routine(
  File "/Users/tjbell1/Research/JWST/ERS/Eureka/ecfs_NIRISS/../src/eureka/S3_data_reduction/niriss_extraction.py", line 434, in optimal_extraction_routine
    es, ev, p = extraction_routine(data, meta, var, spectrum, spectrum_var,
  File "/Users/tjbell1/Research/JWST/ERS/Eureka/ecfs_NIRISS/../src/eureka/S3_data_reduction/niriss_extraction.py", line 518, in extraction_routine
    V = var[i] + np.abs(sky_bkg[i]+(P*reference))/Q
ValueError: operands could not be broadcast together with shapes (256,2048) (142,2048)
```
I believe my hack won't end up working for `per_quad=False`, so that'll be something more to address if you try to get `per_quad=False` working again.


Finally, I don't know how to run your class-based NIRISS code, so I can't test it to see if/where any of the updates I've made may break that code.

taylorbell57 commented:
@afeinstein20, FYI I won't have time to work on the NIRISS files for some time (gotta work on MIRI commissioning data and have a wedding this week). I merged in the updates to the main branch to help save you some time though. Do you have an estimate of when you might be able to get S3 working for NIRISS?

kevin218 commented:
@afeinstein20 I see that you closed this PR yesterday. Does this mean you no longer intend to merge [nirHiss](https://github.com/afeinstein20/nirhiss) with Eureka!?

afeinstein20 commented:
The code in this PR is well outdated at this point, therefore I felt it appropriate to close this one. I've started a new branch with updated code for Niriss data reduction. It's my hope to get all of that in a PR by Friday.
